File: Tracing/Kernels.cs
using Template;
using Cloo;
using System;

namespace clrays
{
    public interface VoidKernel<T> where T : struct
    {
        void Execute(ComputeEventList events);
        OpenCLBuffer<T> GetBuffer();
    }

    public interface ResultKernel<T> where T : struct
    {
        void Execute(ComputeEventList events);
        T[] GetResult();
        OpenCLBuffer<T> GetBuffer();
    }

    public class TraceAaKernel : ResultKernel<float>
    {
        private OpenCLKernel kernel;
        private readonly long[] work;
        private readonly float[] data;
        private readonly OpenCLBuffer<float> buffer;
        private readonly OpenCLBuffer<int> scene_params, tex_params;
        private readonly OpenCLBuffer<float> scene_items;
        private readonly OpenCLBuffer<byte> tex_items;
        private bool dirty;

        public TraceAaKernel(string name, OpenCLProgram program, Scene scene, uint width, uint height, uint AA)
        {
            data = new float[width * height * 3];
            Info.SetTimePoint("Init float buffer");
            buffer = new OpenCLBuffer<float>(program, data);
            Info.SetTimePoint("Copy float buffer to device");
            //make kernel and set args
            kernel = new OpenCLKernel(program, name);
            kernel.SetArgument(0, buffer);
            //set constants
            kernel.SetArgument(1, width);
            kernel.SetArgument(2, height);
            kernel.SetArgument(3, AA);
            Info.SetTimePoint("Init trace kernel");
            //make scene buffers
            var scene_raw = scene.GetBuffers();
            var scene_params_raw = scene.GetParamsBuffer();
            Info.SetTimePoint("Init scene buffers");
            scene_params = new OpenCLBuffer<int>(program, scene_params_raw);
            scene_items = new OpenCLBuffer<float>(program, scene_raw);
            Info.SetTimePoint("Copy scene to device");
            var tex_raw = scene.GetTexturesBuffer();
            var tex_params_raw = scene.GetTextureParamsBuffer();
            Info.SetTimePoint("Init texture buffers");
            tex_params = new OpenCLBuffer<int>(program, tex_params_raw);
            tex_items = new OpenCLBuffer<byte>(program, tex_raw);
            Info.SetTimePoint("Copy textures to device");
            //set arrays
            kernel.SetArgument(4, scene_params);
            kernel.SetArgument(5, scene_items);
            kernel.SetArgument(6, tex_params);
            kernel.SetArgument(7, tex_items);
            //work
            work = new long[] { width * AA, height * AA };
            dirty = false;
            Info.MetaSize = (uint)scene_params_raw.Length +
                            (uint)tex_params_raw.Length;
            Info.MetaSize *= sizeof(int);
            Info.SceneSize = (uint)scene_items.Length * sizeof(float);
            Info.FloatMapSize = (uint)buffer.Length * sizeof(float);
            Info.SetTimePoint("Finish trace kernel");
        }

        public void Update()
        {
            scene_params.CopyToDevice();
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            events.Wait();
            dirty = true;
        }

        public float[] GetResult()
        {
            if (dirty)
                buffer.CopyFromDevice();
            dirty = false;
            return data;
        }

        public OpenCLBuffer<float> GetBuffer()
        {
            return buffer;
        }
    }

    public class TraceKernel : ResultKernel<int>
    {
        private OpenCLKernel kernel;
        private readonly long[] work;
        private readonly int[] data;
        private readonly OpenCLBuffer<int> buffer;
        private readonly OpenCLBuffer<int> scene_params, tex_params;
        private readonly OpenCLBuffer<float> scene_items;
        private readonly OpenCLBuffer<byte> tex_items;
        private bool dirty;

        public TraceKernel(string name, OpenCLProgram program, Scene scene, uint width, uint height)
        {
            data = new int[width * height];
            Info.SetTimePoint("Init int buffer");
            buffer = new OpenCLBuffer<int>(program, data);
            Info.SetTimePoint("Copy int buffer to device");
            //make kernel and set args
            kernel = new OpenCLKernel(program, name);
            kernel.SetArgument(0, buffer);
            //set constants
            kernel.SetArgument(1, width);
            kernel.SetArgument(2, height);
            Info.SetTimePoint("Init trace kernel");
            //make scene buffers
            var scene_raw = scene.GetBuffers();
            var scene_params_raw = scene.GetParamsBuffer();
            Info.SetTimePoint("Init scene buffers");
            scene_params = new OpenCLBuffer<int>(program, scene_params_raw);
            scene_items = new OpenCLBuffer<float>(program, scene_raw);
            Info.SetTimePoint("Copy scene to device");
            var tex_raw = scene.GetTexturesBuffer();
            var tex_params_raw = scene.GetTextureParamsBuffer();
            Info.SetTimePoint("Init texture buffers");
            tex_params = new OpenCLBuffer<int>(program, tex_params_raw);
            tex_items = new OpenCLBuffer<byte>(program, tex_raw);
            Info.SetTimePoint("Copy textures to device");
            //set arrays
            kernel.SetArgument(3, scene_params);
            kernel.SetArgument(4, scene_items);
            kernel.SetArgument(5, tex_params);
            kernel.SetArgument(6, tex_items);
            //work
            work = new long[] { width, height };
            dirty = false;
            Info.MetaSize = (uint)scene_params_raw.Length +
                            (uint)tex_params_raw.Length;
            Info.MetaSize *= sizeof(int);
            Info.SceneSize = (uint)scene_items.Length * sizeof(float);
            Info.IntMapSize = (uint)buffer.Length * sizeof(int);
            Info.SetTimePoint("Finish trace kernel");
        }

        public void Update()
        {
            scene_params.CopyToDevice();
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            events.Wait();
            dirty = true;
        }

        public int[] GetResult()
        {
            if (dirty)
                buffer.CopyFromDevice();
            dirty = false;
            return data;
        }

        public OpenCLBuffer<int> GetBuffer()
        {
            return buffer;
        }
    }

    public class ClearKernel : VoidKernel<float>
    {
        private OpenCLKernel kernel;
        private readonly OpenCLBuffer<float> buffer;
        private readonly long[] work;

        public ClearKernel(string name, OpenCLProgram program, OpenCLBuffer<float> buffer, uint width, uint height)
        {
            kernel = new OpenCLKernel(program, name);
            kernel.SetArgument(0, buffer);
            kernel.SetArgument(1, width);
            kernel.SetArgument(2, height);
            work = new long[] { width, height };
            this.buffer = buffer;
            Info.SetTimePoint("Init clear kernel");
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            events.Wait();
        }

        public OpenCLBuffer<float> GetBuffer()
        {
            return buffer;
        }
    }

    public class ImageKernel : ResultKernel<int>
    {
        private OpenCLKernel kernel;
        private readonly long[] work;
        private readonly int[] data;
        private OpenCLBuffer<int> buffer;
        private bool dirty;

        public ImageKernel(string name, OpenCLProgram program, OpenCLBuffer<float> input, uint width, uint height)
        {
            data = new int[width * height];
            Info.SetTimePoint("Init int buffer");
            buffer = new OpenCLBuffer<int>(program, data);
            Info.SetTimePoint("Copy int buffer to device");
            kernel = new OpenCLKernel(program, name);
            kernel.SetArgument(0, input);
            kernel.SetArgument(1, buffer);
            kernel.SetArgument(2, width);
            kernel.SetArgument(3, height);
            work = new long[] { width, height };
            dirty = false;
            Info.IntMapSize = (uint)data.Length * sizeof(int);
            Info.SetTimePoint("Finish image kernel");
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            events.Wait();
            dirty = true;
        }

        public int[] GetResult()
        {
            if (dirty)
                buffer.CopyFromDevice();
            dirty = false;
            return data;
        }

        public OpenCLBuffer<int> GetBuffer()
        {
            return buffer;
        }
    }
}

File: clrays-rs/src/kernels.rs
use ocl::{Kernel,Program,Queue};
use std::rc::Rc;
use crate::cl_helpers::{ClBuffer};
use crate::scene::Scene;
use crate::info::Info;

pub trait VoidKernel<T: ocl::OclPrm>{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>;
    fn get_buffer(&self) -> &ClBuffer<T>;
}

pub trait ResultKernel<T: ocl::OclPrm>{
    fn get_result(&mut self, queue: &Queue) -> Result<&[T],ocl::Error>;
}

pub struct ClearKernel{
    kernel: Kernel,
    buffer: Rc<ClBuffer<f32>>,
}

impl ClearKernel{
    pub fn new(name: &str, (w,h): (u32,u32), program: &Program, queue: &Queue, buffer: Rc<ClBuffer<f32>>) -> Result<Self,ocl::Error>{
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(buffer.get_ocl_buffer())
        .arg(&10f32)
        .build());
        Result::Ok(Self{ kernel,buffer })
    }
}

impl VoidKernel<f32> for ClearKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => Ok(()),
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<f32>{
        &self.buffer
    }
}

pub struct ImageKernel{
    kernel: Kernel,
    buffer: ClBuffer<i32>,
    dirty: bool,
}

impl ImageKernel{
    pub fn new(name: &str, (w,h): (usize,usize), program: &Program, queue: &Queue, input: &ClBuffer<f32>) -> Result<Self,ocl::Error>{
        let dirty = false;
        let buffer = match ClBuffer::<i32>::new(queue, w * h, 0){
            Ok(x) => x,
            Err(e) => return Err(e),
        };
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(input.get_ocl_buffer())
        .arg(buffer.get_ocl_buffer())
        .arg(w as u32)
        .arg(h as u32)
        .build());
        Ok(Self{ buffer,kernel,dirty })
    }
}

impl VoidKernel<i32> for ImageKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => { self.dirty = true; Ok(()) },
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<i32>{
        &self.buffer
    }
}

impl ResultKernel<i32> for ImageKernel{
    fn get_result(&mut self, queue: &Queue) -> Result<&[i32],ocl::Error>{
        if self.dirty {
            match self.buffer.download(queue){
                Ok(_) => {},
                Err(e) => return Err(e),
            }
        }
        self.dirty = false;
        Ok(self.buffer.get_slice())
    }
}

pub struct TraceKernel{
    kernel: Kernel,
    dirty: bool,
    buffer: ClBuffer<i32>,
    scene_params: ClBuffer<i32>,
    scene_items: ClBuffer<f32>,
    tex_params: ClBuffer<i32>,
    tex_items: ClBuffer<u8>,
    res: (u32,u32),
}

impl TraceKernel{
    pub fn new(name: &str, (w,h): (u32,u32), program: &Program, queue: &Queue, scene: &mut Scene, info: &mut Info) -> Result<Self, ocl::Error>{
        info.set_time_point("Start constructing TraceKernel");
        let dirty = false;
        let buffer = unpack!(ClBuffer::<i32>::new(queue, w as usize * h as usize, 0));
        info.set_time_point("Build int frame buffer");
        let scene_raw = scene.get_buffers();
        let scene_params_raw = scene.get_params_buffer();
        info.set_time_point("Build scene data");
        let scene_params = unpack!(ClBuffer::from(queue, scene_params_raw));
        let scene_items = unpack!(ClBuffer::from(queue, scene_raw));
        info.set_time_point("Build scene buffers");
        let tex_raw = scene.get_textures_buffer();
        let tex_params_raw = scene.get_texture_params_buffer();
        info.set_time_point("Build texture data");
        let tex_params = unpack!(ClBuffer::from(queue, tex_params_raw));
        let tex_items = unpack!(ClBuffer::from(queue, tex_raw));
        info.set_time_point("Build texture buffers");
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(buffer.get_ocl_buffer())
        .arg(w as u32)
        .arg(h as u32)
        .arg(scene_params.get_ocl_buffer())
        .arg(scene_items.get_ocl_buffer())
        .arg(tex_params.get_ocl_buffer())
        .arg(tex_items.get_ocl_buffer())
        .build());
        info.set_time_point("Create kernel");
        Ok(Self{ kernel, dirty, buffer, scene_params, scene_items, tex_params, tex_items, res: (w,h) })
    }

    /*pub fn update(&mut self, queue: &Queue) -> Result<(),ocl::Error>{
        match self.scene_params.upload(queue){
            Ok(_) => {}, Err(e) => return Err(e),
        }
        match self.scene_items.upload(queue){
            Ok(_) => {}, Err(e) => return Err(e),
        }
        match self.tex_params.upload(queue){
            Ok(_) => {}, Err(e) => return Err(e),
        }
        match self.tex_items.upload(queue){
            Ok(_) => {}, Err(e) => return Err(e),
        }
        Ok(())
    }*/

    pub fn get_res(&self) -> (u32,u32){
        self.res
    }
}

impl VoidKernel<i32> for TraceKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => { self.dirty = true; Ok(()) },
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<i32>{
        &self.buffer
    }
}

impl ResultKernel<i32> for TraceKernel{
    fn get_result(&mut self, queue: &Queue) -> Result<&[i32],ocl::Error>{
        if self.dirty {
            match self.buffer.download(queue){
                Ok(_) => {},
                Err(e) => return Err(e),
            }
        }
        self.dirty = false;
        Ok(self.buffer.get_slice())
    }
}
